Sources/NIO/EventLoopFuture+Combine.swift

```diff
@@ -0,0 +1,134 @@
+
+#if canImport(Combine)
+import Combine
+
+
+// MARK:- Obtain a publisher from an EventLoopFuture
+@available(OSX 10.15, iOS 13.0, tvOS 13.0, watchOS 6.0, *)
+extension EventLoopFuture {
+    /// Obtain a Combine `Publisher` for this `EventLoopFuture`.
+    @inlinable
+    public var publisher: EventLoopFuture.Publisher {
+        return Publisher(self)
+    }
+}
+
+
+// MARK:- Definition of EventLoopFuture.Publisher
+@available(OSX 10.15, iOS 13.0, tvOS 13.0, watchOS 6.0, *)
+extension EventLoopFuture {
+    /// A `Publisher` that publishes the results of an `EventLoopFuture`.
+    ///
+    /// An `EventLoopFuture.Publisher` is a Combine `Publisher` that will
+    /// publish exactly once, either a value or an error. This publication will
+    /// occur when the `EventLoopFuture` to which it corresponds completes.
+    public struct Publisher {
+        /// The `EventLoopFuture` whose result this `Publisher` publishes.
+        @usableFromInline
+        internal var future: EventLoopFuture<Value>
+
+        @inlinable
+        internal init(_ future: EventLoopFuture<Value>) {
+            self.future = future
+        }
+    }
+}
+
+
+// MARK:- Definition of EventLoopFuture.Publisher.Subscription
+@available(OSX 10.15, iOS 13.0, tvOS 13.0, watchOS 6.0, *)
+extension EventLoopFuture.Publisher {
+    /// A `Subscription` to an `EventLoopFuture.Publisher`.
+    ///
+    /// This object manages actually attaching the callback to the future at the point where
+    /// data is demanded.
+    @usableFromInline
+    internal class Subscription<S: Subscriber> where Failure == S.Failure, Output == S.Input {
+        /// The `EventLoopFuture` that the subscriber is subscribing to the result of.
+        @usableFromInline
+        internal var future: EventLoopFuture<Value>
+
+        /// The subscriber that subscribed to this future.
+        @usableFromInline
+        internal var subscriber: S?
+
+        @inlinable
+        init(future: EventLoopFuture<Value>, subscriber: S) {
+            self.future = future
+            self.subscriber = subscriber
+        }
+    }
+}
+
+
+// MARK:- Publisher conformance
+@available(OSX 10.15, iOS 13.0, tvOS 13.0, watchOS 6.0, *)
+extension EventLoopFuture.Publisher: Publisher {
+    public typealias Output = Value
+    public typealias Failure = Error
+
+    @inlinable
+    public func receive<S>(subscriber: S) where S : Subscriber, Failure == S.Failure, Output == S.Input {
+        subscriber.receive(subscription: EventLoopFuture.Publisher.Subscription(future: self.future, subscriber: subscriber))
+    }
+}
+
+
+// MARK:- Subscription conformance
+@available(OSX 10.15, iOS 13.0, tvOS 13.0, watchOS 6.0, *)
+extension EventLoopFuture.Publisher.Subscription: Subscription {
+    @inlinable
+    func request(_ demand: Subscribers.Demand) {
+        // This is unlikely, but double check that this is actually for non-zero demand.
+        if case .max(let limit) = demand, limit <= 0 {
+            return
+        }
```

Comment:
Always good to keep such checks 👍 
The RS thing to do is actually to signal an error -- there's no situation where negative or zero demand makes sense, so you likely got a "somehow wrong" downstream so best to let them know:

In RS it's rule 3.9: https://github.com/reactive-streams/reactive-streams-jvm#3.9

While the Subscription is not cancelled, Subscription.request(long n) MUST signal onError with a java.lang.IllegalArgumentException if the argument is <= 0. The cause message SHOULD explain that non-positive request signals are illegal.

Combine did not really publish or document if it follows all those rules, but this is a good one to apply I believe.

Reply:
Well the rule must be at least slightly different, because Combine provides [`Demand.none`](https://developer.apple.com/documentation/combine/subscribers/demand/3213707-none) as a helper. I presume this is in place to facilitate the fact that demand can be requested in the return value of [`receive(_:)`](https://developer.apple.com/documentation/combine/subscriber/3213653-receive).

The question is whether the rule is that zero demand may be returned from `receive(_:)` but not passed to `request`, or whether zero demand is allowed in both cases. @parkera, do you have an insight here?

Reply:
You can return none here.

Reply:
All of our operator implementations assert in request if passed demand <= 0.

Reply:
Right, returning `.none` is one thing, actually invoking upstream's `request(0)` is another IMHO, so one should not tolerate zero demand being signalled into an receive. I'd argue even noop or or other implementations should check `demand > 0`. IMHO best and most useful to signal downstream an error if someone does so, rather than precondition but definitely not silently ignore it..

Makes the returning demand semantics a bit weird, but I know they're useful for performance optimization reasons so so be it.

Reply:
I guess you don't want to signal the error downstream since many impls can be `Failure = Never` thus the decision to make it a fault? That's fair enough to be honest, makes sense given that we only want to signal errors when types allow to after all 🤔

Reply:
Right - unlike Java we can't always assume error, or even the type of the error.